src/pip/_internal/index.py

```diff
@@ -189,14 +190,26 @@ def _get_html_page(link, session=None):
         return None
 
     # Tack index.html onto file:// URLs that point to directories
-    scheme, _, path, _, _, _ = urllib_parse.urlparse(url)
+    scheme, netloc, path, _, _, _ = urllib_parse.urlparse(url)
     if (scheme == 'file' and os.path.isdir(urllib_request.url2pathname(path))):
         # add trailing slash if not present so urljoin doesn't trim
         # final segment
         if not url.endswith('/'):
             url += '/'
         url = urllib_parse.urljoin(url, 'index.html')
         logger.debug(' file: URL is directory, getting %s', url)
+    if (
+        scheme != 'file' and
+        unresponsive_hosts and
+        unresponsive_hosts.is_tracked(netloc)
+    ):
```

Comment:
this would remain the same if we use a set, but otherwise agreed :)

Reply:
One reason I wouldn't put the logic inside the class, is because it is still dependent on the connection error being caught in order to add the host to track. I think to get to a fully self-contained behaviour, it would have to (as per one of your suggestions) involve wrapping the session / session.request entirely.